Nearest neighbors

Riad-Quadratic opened the pull request:
This PR initiate a first version of K nearest neighbors using FHE. 

It includes a streamlit demo app, that is also deployed on HuggingFace:
https://huggingface.co/spaces/Quadratic-Labs/PrivateNearestNeighbors-FHE

bcm-at-zama commented:
Hello @Riad-Quadratic . Thanks a lot for your commit. 

Let me put some review that I did some time ago, but missed to push, sorry for that. 

```
## I had a look to https://github.com/Quadratic-Labs/fhe-tutorials.git, 

commit 391225d6601ddf7de2bc7a46e6952859c4e7af98 (HEAD -> main, origin/main, origin/HEAD)

Merge pull request #15 from Quadratic-Labs/travel_routing

INI: Travel routing

)

In terms of Getting-Started.ipynb:
- I would rename it with something more explicit: like TravelRoutine.ipynb
- I am not sure the use of overpass-turbo is easy for everyone: could we have something simpler, or maybe a .sh to install automatically what needs to be installed, with instructions in the README.md
- It would also certainly be useful to explain a bit more what https://www.qgis.org/fr/site/ is
- Then, it's very important to explain the FHE related points:
        - why do we need privacy here
        - what do you protect in your current implementation (the src, dest, travel?)
        - which function is computed in FHE ("route", right?); by the way, a more explicit function name would be useful, like "compute_journey_in_fhe", or something like that
        - how you deal with FHE in this function? I think here, it's a gigantic table lookup, so you would explain the limits of the approach, like it can't deal with two many nodes (give the limit)
        - also, explain what happens in shortest_path; in particular, I am intrigued by the for-loop, with the individual run/decrypt in there. And the "Careful: breaking early could lead to information leak" does seem frightning a bit. 
        - for things which are work in progress, I would mention it, such that readers know it will be improved; like "WIP: currently, this part is done insecurely, we'll improve this in the next release by doing THIS or THAT"
- Certainly, you should give a better / longer inputset when you compile

    circuit = route.compile([(0, N_NODES - 1), (N_NODES - 1, 0)])

Here it works because you just have a TLU, but when/if you have more complicated function, you'll be happy to have a better / longer inputset
- I feel the benchmark section is a bit too short: I would love to see
    - information on the machine you used
    - benchmarks for different N_NODES
    - also, the benchmarks should depend on the src/dest, no? (since shortest_path will have more or less 'run')
- "shortest_path(22, 28)" is not sufficient for people to understand; could you print the map somehow, then show the 22 and 28 src / dest nodes, and then show the path and we would see the 23 / 2 intermediate nodes
- in your final map, could you highlight 22, 28, 23 and 2 nodes?
- we can't see the 'm' at the end, unless we trust the notebook, can we fix that?



## After I had a look to https://github.com/Quadratic-Labs/fhe-tutorials/tree/nearest_neighbors

commit 6389f477efe010bad737f88f7cda79af99ccbba7 (HEAD -> nearest_neighbors, origin/nearest_neighbors)

Merge pull request #15 from Riad-Quadratic/nearest_neighbors

REFACTOR: clean code
)

- Getting-Started.ipynb has disappeared, expected?
- missing README
- the name of arg_selection function is a bit unclear; and if you mention "relu packed", should we check that x > 0 as well?
- idx = arg_selection(2 * (this_idx - that_idx) + is_positive(diff)):

    I like this trick! nice, explain it

- explain how swap works, by eg taking an example when this > that, and another example where this < that. You would show the intermediate values

    if this > that:

        diff = this_dist - that_dist > 0
        idx = arg_selection(2 * (this_idx - that_idx) + is_positive(diff)) = this_idx - that_idx
        dist = relu(diff) = diff

        idx_min = thas_idx
        idx_max = this_idx
        dist_min = that_dist
        dist_max = this_dist

    else:

        diff = 0
        idx = 0
        dist = 0

        everything else unchanged

- how many PBS do you do in swap? you should explain + show the MLIR, to explain. You could explain how you improved it, in terms of speed

    if I am not wrong, you need 2 PBS with "diff" as input and 1 PBS with an input of the size of 2*N_PTS+1
    what's the range of diff?

- could you describe also which algorithm you KNN is based on? maybe put links to wikipedia if it's a classical algorithm. 
- I guess in your KNN, the 'k' is constant; would you write it somewhere, please? it's k = 2, right? maybe you would then rename your function like knn_k_equal_2 or something like that
- your inputset is too small
- it would be interesting to show the MLIR + compilation time; it's known that Concrete has difficulties with too large MLIR for now
- could you add tests to check that:
    - swap does what it's supposed to do
    - knn does what it's supposed to do
    - nearest does what it's supposed to do
- in your benchmark, could you show results with different number of points N_PTS please? what's the max N_PTS that you can deal with (compile, execute)
- how does your KNN compare with the one in Concrete ML? https://github.com/zama-ai/concrete-ml/blob/main/docs/advanced_examples/KNearestNeighbors.ipynb; you would need to say how your KNN is better / different
- ok, looking at your HF space, it looks it's no more related to TravelRoutine but to "nearest restaurants": am I on the right bounty https://github.com/zama-ai/fhe-tutorials/pull/16? I am a bit confused with the different repositories, I'll check with AlexQ when he is back
- things look very long on HF: what's the execution time for key generation, and for the KNN, typically?
```

Riad-Quadratic commented:
Hello @bcm-at-zama, 

I adressed most of the points related to KNN from your comment. As for:

> - ok, looking at your HF space, it looks it's no more related to TravelRoutine but to "nearest restaurants": am I on the right bounty https://github.com/zama-ai/fhe-tutorials/pull/16? I am a bit confused with the different repositories, I'll check with AlexQ when he is back

You're correct, this HF space is related to a second use case "KNN". The TravelRouting one is still available in the [TravelRouting](https://github.com/zama-ai/fhe-tutorials/tree/main/TravelRouting) folder. The two demos are in two different folders. Let me know if it's unclear.

bcm-at-zama commented:
Thanks for the update, @Riad-Quadratic . I had another look, indeed a few things were done but I can still find questions / points which were not addressed, in my understanding. Could you tell me? Let me name points such that you can re-use my naming if you want to answer to a specific point.

bcm-at-zama commented:
More important:
- A: how does your KNN compare with the one in Concrete ML? https://github.com/zama-ai/concrete-ml/blob/main/docs/advanced_examples/KNearestNeighbors.ipynb; you would need to say how your KNN is better / different; if it's not better, that's an issue; in https://github.com/zama-ai/bounty-and-grant-program/issues/7, we had spoken about 'Djikstra, Bellman-Ford, A*, and many others.'
- B: you need to explain in the notebook how you deal with FHE in this function. I think here, it's a gigantic table lookup, so you would explain the limits of the approach, like it can't deal with two many nodes (give the limit)
- C: also, explain what happens in nearest; in particular, I am intrigued by the for-loop, with the individual run/decrypt in there. Did you remove comments (WIP etc?)? But did you also change the code? Is it really secure, ie does not leak anything?

Important:
- D: fix the unit test for swap: 

        print(f"New index of this is {that_idx}")
        print(f"New value of this is {that_dist}")

        is wrong, you want 'that' in the string 

- E: could you add tests to check that:
    - knn does what it's supposed to do
    - nearest does what it's supposed to do

Other:
- F: why do you write (this_idx - that_idx) + (this_idx - that_idx), instead of 2 * (this_idx - that_idx), in the code and .md?
- G: rename Getting-Started.ipynb it with something more explicit like TravelRoutine.ipynb
- H: Certainly, you should give a better / longer inputset when you compile
- I: why is there 2 first location, 2 second location, 2 third location? (by opposition to one for each

Typos:
- demostrate
- the K nearest restaurant ->  the K nearest restaurants
- perfoms
- swapped with 2 p
- everythin

J: Things are very slow on HF, it gives a bad user experience, I don't know what we can do here.

Riad-Quadratic commented:
Hello @bcm-at-zama, 

Thank you for your comment. I updated the code to take into account most of your points.

> A: how does your KNN compare with the one in Concrete ML? https://github.com/zama-ai/concrete-ml/blob/main/docs/advanced_examples/KNearestNeighbors.ipynb; you would need to say how your KNN is better / different; if it's not better, that's an issue; 

As far as I understand, concrete ML's KNN is a learning algorithm that can be used for classification or regression. The KNN proposed in this demo gives the points of interest closest to a given input. However, I have adapted the concrete ML KNN to achieve the same objective. I've added a folder called "comparison" where you can find a running example for both algorithms. They both give the same results. However, the execution time of the concrete ML algorithm is about 50 times slower. I tried the concrete ML solution for different values of nbits. When it comes to 3, 4 or 5, the execution crashes, whereas for nbits=2, it gives innacurate results. For 6 and more, it works but it takes about 20 minutes to find the nearest neighbours. Please let me know if I've missed anything.

> in https://github.com/zama-ai/bounty-and-grant-program/issues/7, we had spoken about 'Djikstra, Bellman-Ford, A*, and many others.'

The link you provided is a discussion about travel routing. This code is about another use case.

> B: you need to explain in the notebook how you deal with FHE in this function. I think here, it's a gigantic table lookup, so you would explain the limits of the approach, like it can't deal with two many nodes (give the limit)

Yes, it is indeed a large table lookup. This point will be dealt with in depth later. For now, there's a paragraph at the end of the README showing the execution time as the number of nodes increases.

> C: also, explain what happens in nearest; in particular, I am intrigued by the for-loop, with the individual run/decrypt in there. Did you remove comments (WIP etc?)? But did you also change the code? Is it really secure, ie does not leak anything?

I can’t see which loop are you referring to? Are you referring the travel routing notebook? If not, can you please provide a snippet of that code. 

 
>     D: fix the unit test for swap:
> 
>       print(f"New index of this is {that_idx}")
>       print(f"New value of this is {that_dist}")
> 
>       is wrong, you want 'that' in the string 

Done

>     E: could you add tests to check that:
>         knn does what it's supposed to do
>         nearest does what it's supposed to do


Nearest is tested in the last cell of the getting_started.ipynb notebook (now it’s called knn.ipynb). The test shows that for three different locations, the results are indeed the closest points. The test of KNN is implicit as the nearest function is just a call to KNN wrapped with encryption/decryption. However, I added a new cell that asserts that nearest and KNN return the same outputs for the same inputs.


>     F: why do you write (this_idx - that_idx) + (this_idx - that_idx), instead of 2 * (this_idx - that_idx), in the code and .md?

I naively thought that multiplications were more expensive than additions. I've corrected that.

>     G: rename Getting-Started.ipynb it with something more explicit like TravelRoutine.ipynb

Done, now it's knn.ipynb

>     H: Certainly, you should give a better / longer inputset when you compile

I changed it, but could you, please, give me an example of how large the input set should be?

>     I: why is there 2 first location, 2 second location, 2 third location? (by opposition to one for each

I'm not sure about this point. There's only one for each. You're probably referring to the coordinates because here we pass each coordinate separately. So, first_location[0] represents the first coordinate while first_location[1] is the second.

> Typos:
> 
>     demostrate
>     the K nearest restaurant -> the K nearest restaurants
>     perfoms
>     swapped with 2 p
>     everythin

Done

> J: Things are very slow on HF, it gives a bad user experience, I don't know what we can do here.

Unfortunately, this is the best we could do for this first version. We are considering optimizations that may come later.

Let me know if there is something unclear,

bcm-at-zama commented:
Hello @Riad-Quadratic 

For A: thanks a lot, we'll have a look to your KNN then, it's https://github.com/zama-ai/concrete-ml-internal/issues/4288

For zama-ai/bounty-and-grant-program#7: indeed my bad, sorry

For B: "Yes, it is indeed a large table lookup. This point will be dealt with in depth later.". It's not clear to me when is this "later", we'll speak about it internally

For C: ok indeed, I was confusing here as well with travel rountine, my bad

For D: thanks

For E: I would have done much more extensive tests:
    - with random this_idx, this_dist, that_idx, that_dist and call to swap
    - with random x, y and call to knn
    - with random x, y and call to nearest
and with assertions that the results are as expected

For F: multiplication by 2 or addition are equivalent

For G: thanks

For H: better to use random inputset where data has the right range. So, something like

    max_index = len(points_array)
    inputset = [(np.random.randint(max_index), np.random.randint(max_index)) for _ in range(100)]

For I: ok so maybe it would be clearer with

    first_location = (2, 1)
    closest_to_first_location = nearest(*first_location)

and so on? With that, at least to me, it's clearer that we speak about Point (2, 1)

bcm-at-zama commented:
Let me discuss with the rest of the team (in two weeks, because of vacations) for the next steps for the bounty. Certainly it's time to merge but I'd like to check with them.
